diff --git a/WinHTTrack/Shell.cpp b/WinHTTrack/Shell.cpp
index 1eb07d8..f4b1252 100755
--- a/WinHTTrack/Shell.cpp
+++ b/WinHTTrack/Shell.cpp
@@ -2478,7 +2478,7 @@ void Write_profile(CString path, int load_path) {
 		MyWriteProfileInt(path, strSection, "StoreAllInCache", maintab->m_option9.m_Cache2);
 		MyWriteProfileInt(path, strSection, "LogType", maintab->m_option9.m_logtype);
 		MyWriteProfileInt(path, strSection, "UseHTTPProxyForFTP", maintab->m_option10.m_ftpprox);
-		MyWriteProfileInt(path, strSection, "UseHTTPProxyForHTTPS", maintab->m_option10.m_ftpprox);
+		MyWriteProfileInt(path, strSection, "UseHTTPProxyForHTTPS", maintab->m_option10.m_sslproxy);
 
 		// menus
 		MyWriteProfileInt(path, strSection, "Build", maintab->m_option2.m_build);
@@ -2838,7 +2838,7 @@ void Read_profile(CString path, int load_path) {
 	maintab->m_option10.m_proxy = MyGetProfileString(path, strSection, "Proxy");
 	maintab->m_option10.m_port = MyGetProfileString(path, strSection, "Port");
 	maintab->m_option10.m_ftpprox = MyGetProfileInt(path, strSection, "UseHTTPProxyForFTP", 1);
-	maintab->m_option10.m_ftpprox = MyGetProfileInt(path, strSection, "UseHTTPProxyForHTTPS", 1);
+	maintab->m_option10.m_sslproxy = MyGetProfileInt(path, strSection, "UseHTTPProxyForHTTPS", 1);
 	//
 	maintab->m_option5.m_depth = MyGetProfileString(path, strSection, "Depth");
 	maintab->m_option5.m_depth2 = MyGetProfileString(path, strSection, "ExtDepth");
diff --git a/httrack b/httrack
--- a/httrack
+++ b/httrack
@@ -1 +1 @@
-Subproject commit ed9e25ceca944fd92be5c5c324f4c934327bd0c5
+Subproject commit ed9e25ceca944fd92be5c5c324f4c934327bd0c5-dirty
