diff --git a/src/htslib.c b/src/htslib.c
index e900a3c..a17c1af 100644
--- a/src/htslib.c
+++ b/src/htslib.c
@@ -3424,7 +3424,7 @@ HTSEXT_API const char* jump_normalized_const(const char* source) {
 	if (strcmp(source, "file://") == 0)
 		return source;
 	source = jump_identification_const(source);
-	if (strfield(source, "www") && source[3] != '\0') {
+	/*if (strfield(source, "www") && source[3] != '\0') {
 		if (source[3] == '.') {     // www.foo.com -> foo.com
 			source += 4;
 		} else {                    // www-4.foo.com -> foo.com
@@ -3436,7 +3436,7 @@ HTSEXT_API const char* jump_normalized_const(const char* source) {
 				source = a + 1;
 			}
 		}
-	}
+	}*/
 	return source;
 }
 
@@ -3914,8 +3914,10 @@ HTSEXT_API size_t x_escape_flags(const char* const s, char* const dest, const si
 		} else {
 			if (flags & ENCODE_PARAMS_AS_HTML_ENTITIES) {
 				ADD_CHAR('&');
-				ADD_CHAR('#');
-				ADD_CHAR('x');
+				if (c != '&') {
+					ADD_CHAR('#');
+					ADD_CHAR('x');
+				}
 			} else
 				ADD_CHAR('%');
 			if (c != '&' || (flags & ENCODE_PARAMS_AS_HTML_ENTITIES) == FALSE) {
